### ISSUE TITLE
Add support for mutations outside callbacks

### ISSUE THREAD

**Issue by JamesPHoughton** [empiricaly/empirica#266]
To trigger a specific callback (and set a value that I'll use in the callback and downstream), I'm using a variable that I can set when I want the callback to fire:
```js
// server/src/callbacks.js
Empirica.on("batch", "status", (ctx, { batch, status }) => {
  if (status === "running") {
      console.log("value set at:", Date.now());
      batch.set("mySweetKey", true);
    }
  }

Empirica.on("batch", "mySweetKey", () => {
  console.log("callback fired at:", Date.now();
}
```

This results in the time between the value being set and the callback being fired being very small (9 ms in this example, there are other server events happening in between).
```bash
deliberation-empirica-empirica-1  | [server] value set at: 1678730643209
... (other log messages)
deliberation-empirica-empirica-1  | [server] callback fired at: 1678730643228
```

However, I want my custom callback to fire 10 seconds after the batch status callback, so I use a timer:

```js
// server/src/callbacks.js
Empirica.on("batch", "status", (ctx, { batch, status }) => {
  if (status === "running") {
      setTimeout(() => {
        console.log("value set at:", Date.now());
        batch.set("mySweetKey", true);
      }, 10000);
    }
  }

Empirica.on("batch", "mySweetKey", () => {
  console.log("callback fired at:", Date.now();
}

```
(In this simplified example I could have put the timer inside the callback, or just run a function instead of using a callback, but this is a minimum replicating example of something more complex.)

What happens now is that the time between the value being set and the callback firing can be arbitrarily large (~17 seconds in the example below). It seems that the callback only fires after some other event happens on the server, as the server sits after printing line 1 below, and then waits for 17 seconds, and as soon as line 2 is printed, line 3 is printed below it:

```bash
1. deliberation-empirica-empirica-1  | [server] value set at: 1678731349006
2. deliberation-empirica-empirica-1  | [server] stage transition check
3. deliberation-empirica-empirica-1  | [server] callback fired at: 1678731406222
```

This may have nothing strictly to do with the timer, but just arise because without the timer, when you set a value on the server there is usually something else happening to help the trigger fire. I would have expected setting the value to trigger the callback without needing some other event to activate it, though. 

# Environment:
- "@empirica/core": "1.0.6",
- Docker build FROM ghcr.io/empiricaly/empirica:build-249

**Comment by npaton** [comment]
Short answer: changing values at later time is not supported. So the `.set` happening in the setTimeout callback will not be trigger when you think it would. It has to do with how we save value at the end of the callback. Since we are not using async, we track all the changes during the callback, then apply then when the callback finishes. If you do something outside of that, we don't apply the changes (or we do actually but not synchronously to the time you use `.set`).

We should be able to support this, and I need to think about it. Maybe wrapping callbacks like the one in setTimeout. Maybe explicitly calling a `flush` function. I don't yet. In the short term, we simply don't have a way to support this I think.

**Comment by JamesPHoughton** [comment]
Is there a way to tap into the existing timing infrastructure inside of Empirica that manages advancing rounds/stages? I don't know how exactly that works, but it would be neat to be able to manually register a timer with a UTC/epoch time, and have Empirica fire a supplied callback at that point. Either a universal set of timers that you could register, or ones specific to a game that would get the game object as a parameter:
```js
function myTimerCallback({game, time}) {
  console.log(`Its ${time}, setting to spooky mode)`)
  game.set("SpookyMode", true)
}

Empirica.on('game', 'start', (ctx, { game, start }) => {
  const midnightTonight = ... // epoch time
  game.registerTimer(midnightTonight, myTimerCallback)
}
```

If we couldn't tap into the existing timing infrastructure, I could imagine also maintaining a queue in parallel to the callbacks queue, and after each callback resolves (or at least once a second) first checking the timer queue for any expired timers, before running the next callback.

**Comment by npaton** [comment]
Added a way to update attributes outside callbacks. Please check out the release notes for more info: https://github.com/empiricaly/empirica/releases/tag/%40empirica%2Fcore%401.7.0

As for the managed timer, I created a separate issue for the: https://github.com/empiricaly/empirica/issues/407
